File: packages/typescript/device-client-common/src/device-host-client.ts
import { closeWebSocketWithTruncateReason, errorify, Instance, stringify, transformAndValidate, WebSocketSpec } from '@dogu-tech/common';
import { DeviceHostUploadFileReceiveMessage, DeviceHostUploadFileSendMessage, isAllowedBrowserName, isAllowedBrowserPlatform, ThirdPartyPathMap } from '@dogu-tech/types';
import { DeviceClientOptions, DeviceCloser, DeviceService, DeviceWebSocket, HostFileUploader } from './bases';
import { DeviceHttpClient } from './device-http-client';
import { DeviceHost } from './specs/http/device-host';
import { DeviceHostDownloadSharedResource } from './specs/ws/device-host/download-shared-resource';
import { DeviceHostEnsureBrowserAndDriver } from './specs/ws/device-host/ensure-browser-and-driver';
import { DeviceHostUploadFile } from './specs/ws/device-host/upload-file';

export class DeviceHostClient extends DeviceHttpClient {
  constructor(deviceService: DeviceService, options?: DeviceClientOptions) {
    super(deviceService, options);
  }

  async getFreePort(excludes?: number[], offset?: number): Promise<number> {
    const response = await this.httpRequest(DeviceHost.getFreePort, new DeviceHost.getFreePort.pathProvider(), {
      excludes,
      offset,
    });
    const { port } = response;
    return port;
  }

  async getPathMap(): Promise<ThirdPartyPathMap> {
    const response = await this.httpRequest(DeviceHost.getPathMap, new DeviceHost.getPathMap.pathProvider());
    const { pathMap } = response;
    return pathMap;
  }

  private async connectWebSocket<SendMessageType, ReceiveMessageType, ReturnType>(
    webSocketSpec: WebSocketSpec<SendMessageType, ReceiveMessageType>,
    onOpen: (deviceWebSocket: DeviceWebSocket) => ReturnType,
    onClose: (code: number, reason: string) => void,
    onMessage: (value: string | Uint8Array, deviceWebSocket: DeviceWebSocket) => void,
  ): Promise<ReturnType> {
    return new Promise((resolve, reject) => {
      let isOpened = false;
      const path = webSocketSpec.path;
      const deviceWebSocket = this.deviceService.connectWebSocket(
        {
          path,
        },
        this.options,
        {
          onOpen() {
            isOpened = true;
            resolve(onOpen(deviceWebSocket));
          },
          onClose(ev) {
            const { code, reason } = ev;
            if (!isOpened) {
              reject(new Error(`Unexpected close: ${code} ${reason}`));
              return;
            }
            onClose(code, reason);
          },
          onMessage(ev) {
            const { value } = ev;
            if (!value) {
              throw new Error(`Unexpected data: ${stringify(ev)}`);
            }
            const { $case } = value;
            if ($case === 'bytesValue') {
              onMessage(value.bytesValue, deviceWebSocket);
            } else if ($case === 'stringValue') {
              onMessage(value.stringValue, deviceWebSocket);
            } else {
              throw new Error(`Unexpected data: ${stringify(value)}`);
            }
          },
        },
      );
    });
  }

  async uploadFile(fileName: string, fileSize: number, onProgress: (offset: number) => void, onComplete: (filePath: string, error?: Error) => void): Promise<HostFileUploader> {
    return this.connectWebSocket(
      DeviceHostUploadFile,
      (deviceWebSocket) => {
        const sendMessage: Instance<typeof DeviceHostUploadFile.sendMessage> = {
          value: {
            $case: 'start',
            start: {
              fileName,
              fileSize,
            },
          },
        };
        deviceWebSocket.send(DeviceHostUploadFileSendMessage.encode(sendMessage).finish());
        return new HostFileUploader(deviceWebSocket);
      },
      (code, reason) => {
        if (code === 1000) {
          return;
        }
        onComplete('', new Error(`Unexpected close: ${code} ${reason}`));
      },
      (value, deviceWebSocket) => {
        if (!(value instanceof Uint8Array)) {
          throw new Error(`Unexpected data: ${stringify(value)}`);
        }

        const receiveMessage = DeviceHostUploadFileReceiveMessage.decode(value);
        if (!receiveMessage.value) {
          throw new Error(`Empty data: ${stringify(receiveMessage)}`);
        }
        const { $case } = receiveMessage.value;
        if ($case === 'inProgress') {
          const { offset } = receiveMessage.value.inProgress;
          onProgress(offset);
        } else if ($case === 'complete') {
          const { filePath } = receiveMessage.value.complete;

          onComplete(filePath);
          deviceWebSocket.close(1000, 'OK');
        }
      },
    );
  }

  async downloadSharedResource(filePath: string, url: string, expectedFileSize: number, headers?: Record<string, string>): Promise<void> {
    const { printable } = this.options;
    return new Promise((resolve, reject) => {
      this.connectWebSocket(
        DeviceHostDownloadSharedResource,
        (deviceWebSocket) => {
          const sendMessage: Instance<typeof DeviceHostDownloadSharedResource.sendMessage> = {
            url,
            headers,
            filePath,
            expectedFileSize,
          };
          deviceWebSocket.send(JSON.stringify(sendMessage));
          return new DeviceCloser(deviceWebSocket);
        },
        (code, reason) => {
          if (code === 1000) {
            resolve();
          } else {
            reject(new Error(`Unexpected close: ${code} ${reason}`));
          }
        },
        (value, deviceWebSocket) => {
          const casted = typeof value === 'string' ? value : Buffer.from(value).toString();
          transformAndValidate(DeviceHostDownloadSharedResource.receiveMessage, JSON.parse(casted))
            .then((message) => {
              const { responseCode, responseHeaders } = message;
              printable.info('Download Complete', {
                filePath,
                responseCode,
                responseHeaders,
              });
            })
            .catch((error) => {
              printable.error('Failed to parse download shared resource message', { error: errorify(error) });
              closeWebSocketWithTruncateReason(deviceWebSocket, 1001, 'download shared resource failed');
            });
        },
      )
        .then((closer) => {
          // noop
        })
        .catch((error) => {
          reject(error);
        });
    });
  }

  async ensureBrowserAndDriver(
    params: {
      browserName: string;
      browserVersion: string;
      browserPlatform: string;
      deviceSerial: string;
    },
    options?: {
      timeoutMilliseconds?: number;
    },
  ): Promise<Instance<typeof DeviceHostEnsureBrowserAndDriver.receiveMessage>> {
    const { printable } = this.options;
    const { browserName, browserVersion, browserPlatform, deviceSerial } = params;
    const timeoutMilliseconds = options?.timeoutMilliseconds ?? 60_000;

    return await new Promise((resolve, reject) => {
      let fulfilled = false;
      let timeout: NodeJS.Timeout | null = setTimeout(() => {
        if (fulfilled) {
          return;
        }
        fulfilled = true;
        reject(new Error(`Ensure browser and driver timeout: ${timeoutMilliseconds}ms`));
      });

      const _resolve = (value: Instance<typeof DeviceHostEnsureBrowserAndDriver.receiveMessage>): void => {
        if (fulfilled) {
          return;
        }
        fulfilled = true;

        if (timeout) {
          clearTimeout(timeout);
          timeout = null;
        }

        resolve(value);
      };

      const _reject = (error: Error): void => {
        if (fulfilled) {
          return;
        }
        fulfilled = true;

        if (timeout) {
          clearTimeout(timeout);
          timeout = null;
        }

        reject(error);
      };

      this.connectWebSocket(
        DeviceHostEnsureBrowserAndDriver,
        (deviceWebSocket) => {
          if (!isAllowedBrowserName(browserName)) {
            _reject(new Error(`Invalid browser name: ${browserName}`));
            return;
          }

          if (!isAllowedBrowserPlatform(browserPlatform)) {
            _reject(new Error(`Invalid browser platform: ${browserPlatform}`));
            return;
          }

          const sendMessage: Instance<typeof DeviceHostEnsureBrowserAndDriver.sendMessage> = {
            browserName,
            requestedBrowserVersion: browserVersion,
            browserPlatform,
            deviceSerial,
          };
          deviceWebSocket.send(JSON.stringify(sendMessage));
          return new DeviceCloser(deviceWebSocket);
        },
        (code, reason) => {
          if (code === 1000) {
            // noop
          } else {
            _reject(new Error(`Unexpected close: ${code} ${reason}`));
          }
        },
        (value, deviceWebSocket) => {
          const casted = typeof value === 'string' ? value : Buffer.from(value).toString();
          transformAndValidate(DeviceHostEnsureBrowserAndDriver.receiveMessage, JSON.parse(casted))
            .then((message) => {
              _resolve(message);
            })
            .catch((error) => {
              printable.error('Failed to ensure browser and driver', { error: errorify(error) });
              closeWebSocketWithTruncateReason(deviceWebSocket, 1001, 'ensure browser and driver failed');
            });
        },
      )
        .then((closer) => {
          // noop
        })
        .catch((error) => {
          _reject(errorify(error));
        });
    });
  }
}

File: packages/python/dogu-device-client/dogu/device/device_host_client.py
from dataclasses import asdict, dataclass
from typing import List, Optional
import requests
from dogu.device.common.device_http_response import DeviceHttpResponse
from dogu.device.logger import create_logger
from dogu.device.common.const import DOGU_DEVICE_AUTHORIZATION_HEADER_KEY


@dataclass(frozen=True)
class GetFreePortQuery:
    # pylint: disable=invalid-name
    excludes: List[int]
    # pylint: disable=invalid-name
    offset: int


@dataclass(frozen=True)
class GetFreePortResponse:
    # pylint: disable=invalid-name
    port: int


@dataclass(frozen=True)
class EnsureBrowserAndDriverOptions:
    # pylint: disable=invalid-name
    browserName: str

    # pylint: disable=invalid-name
    browserPlatform: str

    # pylint: disable=invalid-name
    browserVersion: Optional[str]

    # pylint: disable=invalid-name
    deviceSerial: Optional[str]


# pylint: disable=too-many-instance-attributes
@dataclass(frozen=True)
class EnsureBrowserAndDriverResult:
    # pylint: disable=invalid-name
    browserName: str

    # pylint: disable=invalid-name
    browserPlatform: str

    # pylint: disable=invalid-name
    browserVersion: str

    # pylint: disable=invalid-name
    browserMajorVersion: int

    # pylint: disable=invalid-name
    browserDriverVersion: str

    # pylint: disable=invalid-name
    browserDriverPath: str

    # pylint: disable=invalid-name
    browserPath: Optional[str]

    # pylint: disable=invalid-name
    browserPackageName: Optional[str]

    # pylint: disable=invalid-name
    deviceSerial: Optional[str]


class DeviceHostClient:
    def __init__(self, device_server_url: str, token: str, timeout: int = 10):
        self.device_server_url = (
            device_server_url
            if device_server_url[-1] != "/"
            else device_server_url[:-1]
        )
        self._token = token
        self.timeout = timeout
        self._logger = create_logger(__name__)

    def get_free_port(self, excludes: Optional[List[int]] = None) -> int:
        if not excludes:
            excludes = []
        full_path = f"{self.device_server_url}/device-host/free-port"
        param = GetFreePortQuery(excludes=excludes, offset=0)
        headers = {}
        headers[DOGU_DEVICE_AUTHORIZATION_HEADER_KEY] = self._token

        res = requests.get(
            full_path, json=asdict(param), timeout=self.timeout, headers=headers
        )
        res.raise_for_status()
        device_res = DeviceHttpResponse(res)
        if device_res.error()[0]:
            raise Exception(
                f"DeviceHostClient.get_free_port error: {device_res.error()[1].message}"
            )
        res_obj = device_res.data(GetFreePortResponse)
        return res_obj.port

    def ensure_browser_and_driver(
        self, options: EnsureBrowserAndDriverOptions
    ) -> EnsureBrowserAndDriverResult:
        full_path = f"{self.device_server_url}/device-host/ensure-browser-and-driver"
        headers = {}
        headers[DOGU_DEVICE_AUTHORIZATION_HEADER_KEY] = self._token

        res = requests.post(
            full_path, json=asdict(options), timeout=self.timeout, headers=headers
        )
        res.raise_for_status()
        device_res = DeviceHttpResponse(res)
        if device_res.error()[0]:
            raise Exception(
                f"DeviceHostClient.ensure_browser_and_driver error: {device_res.error()[1].message}"
            )
        res_obj = device_res.data(EnsureBrowserAndDriverResult)
        return res_obj
